client.go

```diff
@@ -415,6 +414,53 @@ func (d *Dialer) DialContext(ctx context.Context, urlStr string, requestHeader h
 	return conn, resp, nil
 }
 
+// Returns an updated proxy dial function, wrapping the passed
+// "netDial" function, or an error if one occurred. If the proxy
+// is HTTPS, then TLS handshake will occur in the returned dial function.
+func (d *Dialer) proxyNetDial(req *http.Request, netDial netDialerFunc) (netDialerFunc, error) {
+	if d.Proxy == nil {
+		return nil, errors.New("proxy url function is nil")
+	}
+	// Generate the proxy URL and validate.
+	proxyURL, err := d.Proxy(req)
+	if err != nil {
+		return nil, err
+	}
+	if proxyURL == nil {
+		return nil, errors.New("generated proxy url is nil")
+	}
+	if proxyURL.Scheme == "http" && d.NetDialTLSContext != nil {
+		return nil, errors.New("dial TLS function set for dialing non-HTTPS proxy")
+	}
+	// For HTTPS proxy, either use TLS dial function set in Dialer (which performs
+	// the TLS handshake within the function), or create a new dial function by wrapping
+	// the passed "netDial" for the connection and using TLSClientConfig for the handshake.
+	if proxyURL.Scheme == "https" {
+		if d.NetDialTLSContext != nil {
+			netDial = d.NetDialTLSContext
```

Comment:
note for self:

- url is http and proxy-url is https, we replace the dialer ... users can specify a custom tls dialer just for the proxy.
- url is https and proxy-url is https, the same tls dialer is used for both

Reply:
> * url is https and proxy-url is https, the same tls dialer is used for both

I do *not* think this is correct. There is only one "dial". It is either to the "proxy" or to the "upstream" server. If there is a proxy, the dial is to the "proxy" (and the TLS handshake must occur in the dial). The "proxy" dials then dials the "upstream".  The second TLS handshake (not dial) to the "upstream" occurs over the previously established connection, and it MUST use the `TLSClientConfig`.